JDK-8256725: Metaspace: better blocktree and binlist asserts

tstuefe opened the pull request:
Hi,

may I have reviews please for this small change. 

To analyze JDK-8256572, I'd like to see more information in asserts for binlist and blocktree.

This patch:

- beefs up assertion messages when verifying binlist and blocktree
- adds a canary to the blocktree node to detect overwriters
- improves the blocktree printing
- adds a gtest death test to test the overwrite detection.

Thanks!

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux aarch64 | Linux arm | Linux ppc64le | Linux s390x | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- | ----- | ----- | ----- | ----- |
| Build | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ⏳ (6/6 running) | ⏳ (2/2 running) | ⏳ (2/2 running) | ⏳ (2/2 running) |
| Test (tier1) |    |     |     |     |  ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (7/7 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8256725](https://bugs.openjdk.java.net/browse/JDK-8256725): Metaspace: better blocktree and binlist asserts


### Reviewers
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - **Reviewer**)
 * [Richard Reingruber](https://openjdk.java.net/census#rrich) (@reinrich - Committer)
 * [Leo Korinth](https://openjdk.java.net/census#lkorinth) (@lkorinth - Committer)


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1339/head:pull/1339`
`$ git checkout pull/1339`

shipilev commented on src/hotspot/share/memory/metaspace/binList.hpp:
```diff
@@ -182,8 +185,11 @@ class BinListImpl {
-      for (Block* b = _blocks[i]; b != NULL; b = b->_next) {
-        assert(b->_word_size == s, "bad block size");
+      int pos = 0;
+      for (Block* b = _blocks[i]; b != NULL; b = b->_next, pos ++) {
```
`pos++` seems to be more in-style here.

shipilev commented on src/hotspot/share/memory/metaspace/blockTree.cpp:
```diff
@@ -36,27 +37,40 @@ namespace metaspace {
 
+#define NODE_FORMAT \
+  "@" PTR_FORMAT \
+  ": canary: " INTPTR_FORMAT \
+  ", parent " PTR_FORMAT \
+  ", left " PTR_FORMAT \
```
So, is this message with ":" (like `canary:`), or without it (like all other fields)?

shipilev commented on src/hotspot/share/memory/metaspace/blockTree.cpp:
```diff
@@ -36,27 +37,40 @@ namespace metaspace {
+  p2i((n) ? (n)->_parent : NULL), \
+  p2i((n) ? (n)->_left : NULL), \
+  p2i((n) ? (n)->_right : NULL), \
+  p2i((n) ? (n)->_next : 0), \
```
This is still pointer, right, judging by `p2i`? So it should return `NULL` in the ternary op?

shipilev commented on src/hotspot/share/memory/metaspace/blockTree.cpp:
```diff
@@ -36,27 +37,40 @@ namespace metaspace {
-// These asserts prints the tree, then asserts
-#define assrt(cond, format, ...) \
+// This assert prints the tree, then stops (generic message)
+#define assrt0(cond, format, ...) \
```
The names get hard to read. I thought it was as typo initially :) Suggestion: `tree_assert`?

shipilev commented on src/hotspot/share/memory/metaspace/blockTree.cpp:
```diff
@@ -70,9 +84,11 @@ struct BlockTree::walkinfo {
   // Traverse the tree and test that all nodes are in the correct order.
 
+  // assert a condition with information about node "n"
+  #define assrt0n(cond, failure_node) assrt0(cond, "Invalid node: " NODE_FORMAT, NODE_FORMAT_ARGS(failure_node))
+
```
Do you think this assert is only usable in this method? I would probably be cleaner to move it next to the assertion it chains with.

shipilev reviewed (commented):
Stylistic drive-by comments.

tstuefe commented:
> Stylistic drive-by comments.

Thanks Aleksey! All valid, all fixed.

coleenp commented:
Can you add the hotspot-dev mailing list to these RFRs?

tstuefe commented:
/label add hotspot-dev

tstuefe commented:
I added some more checks - which should be very quick - for all passed nodes when inserting or removing into the BlockTree. Similar for BinList.

lkorinth commented on src/hotspot/share/memory/metaspace/blockTree.cpp:
```diff
@@ -92,25 +109,28 @@ void BlockTree::verify() const {
-      assrt0(n->_word_size > info.lim1);
-      assrt0(n->_word_size < info.lim2);
+      tree_assert_invalid_node(n->_word_size >= MinWordSize &&
+                               n->_word_size <= chunklevel::MAX_CHUNK_WORD_SIZE, n);
```
I prefer one assert per line (that is, not using `&&` in asserts). In this case it does not matter as much, as we can deduce which sub-expression failed from the tree printout,  but I think it is a better style, because in _general_ you do not want such an assert to fail.

reinrich reviewed (commented):
Just a few minor points. In general looks good.

lkorinth reviewed (changes requested):
Looks good. I like the canary in Node. I have three minor comments you should take a look at.

tstuefe commented:
Richard, Leo,

thanks for reviewing!

I fixed your remarks. Changes in the last commit:
- %d -> %u
- split multiline assert into two asserts
- simplified NODE_FORMAT_ARGS
- When verifying a tree, I have now special handling for unreadable node pointers and node pointers whose canaries are invalid. In the former case I assert right away, in the latter I print the node as hex dump (I considered doing further analysis like "is in metaspace" which would be trivial but this class is earmarked as a potential future generic class so I did not want to add deps to metaspace).
- When an error is detected, the whole tree is printed; now tree printing tippytoes around invalid pointers.

lkorinth reviewed (approved):
Looks good to me.

reinrich reviewed (approved):
Looks good.

tstuefe commented:
/integrate

tstuefe commented:
> @tstuefe This PR has not yet been marked as ready for integration.

What? Why?

tstuefe commented:
/integrate

tstuefe commented:
/integrate

shipilev reviewed (approved):
I guess you miss a formal Reviewer ack.

tstuefe commented:
/integrate
